manta-util/src/collections/vec_deque.rs

```diff
@@ -0,0 +1,175 @@
+    #[inline]
+    pub fn at_level(&self, level: usize) -> &VecDeque<T> {
+        &self.0[level]
+    }
+
+    /// Returns a mutable reference to the [`VecDeque`] at the given `level`.
+    #[inline]
+    pub fn at_level_mut(&mut self, level: usize) -> &mut VecDeque<T> {
+        &mut self.0[level]
+    }
+
+    /// Returns the total number of elements in `self`.
+    #[inline]
+    pub fn len(&self) -> usize {
+        self.0.iter().map(VecDeque::len).sum()
+    }
+
+    /// Returns `true` if `self` has no elements.
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        self.0.iter().all(VecDeque::is_empty)
+    }
+
+    /// Returns a shared reference to the element at the given `index` in `self`.
+    #[inline]
+    pub fn get(&self, mut index: usize) -> Option<&T> {
+        for level in self.0.iter() {
+            match level.get(index) {
+                None => index -= level.len(),
+                item => return item,
+            }
+        }
+        None
+    }
+
+    /// Returns a mutable reference to the element at the given `index` in `self`.
+    #[inline]
+    pub fn get_mut(&mut self, mut index: usize) -> Option<&mut T> {
+        for level in self.0.iter_mut() {
+            let len = level.len();
+            match level.get_mut(index) {
+                None => index -= len,
+                item => return item,
+            }
+        }
+        None
+    }
+
+    /// Returns a shared reference to the first element in the deque.
+    #[inline]
+    pub fn front(&self) -> Option<&T> {
+        self.0.iter().find_map(VecDeque::front)
+    }
+
+    /// Returns a mutable reference to the first element in the deque.
+    #[inline]
+    pub fn front_mut(&mut self) -> Option<&mut T> {
+        self.0.iter_mut().find_map(VecDeque::front_mut)
+    }
+
+    /// Returns `true` if `item` is at the front of the deque.
+    #[inline]
+    pub fn is_front(&self, item: &T) -> bool
+    where
+        T: PartialEq,
+    {
+        self.is_front_with(item, PartialEq::eq)
+    }
+
+    /// Returns `true` if `item` is at the front of the deque using `eq` to compare elements of type
+    /// `T`.
+    #[inline]
+    pub fn is_front_with<F>(&self, item: &T, mut eq: F) -> bool
+    where
+        F: FnMut(&T, &T) -> bool,
+    {
+        if let Some(front) = self.front() {
+            eq(front, item)
+        } else {
+            false
+        }
+    }
+
+    /// Finds the position of `item` assuming it was inserted at the given `level`.
+    #[inline]
+    pub fn position(&self, level: usize, item: &T) -> Option<usize>
+    where
+        T: PartialEq,
+    {
+        self.position_with(level, item, PartialEq::eq)
+    }
+
+    /// Finds the position of `item` assuming it was inserted at the given `level` using `eq` to
+    /// compare elements of type `T`.
+    #[inline]
+    pub fn position_with<F>(&self, level: usize, item: &T, mut eq: F) -> Option<usize>
+    where
+        F: FnMut(&T, &T) -> bool,
```

Comment:
Same as above.

Reply:
`position` function requires FnMut(..), so we still need `FnMut` here.
`core::iter::traits::iterator::Iterator
pub fn position<P>(&mut self, predicate: P) -> Option<usize>
where
    Self: Sized,
    P: FnMut(Self::Item) -> bool,`